File: algorithms/c++/438-find-all-anagrams-in-a-string.cpp
// https://leetcode.com/problems/find-all-anagrams-in-a-string/
#include <bits/stdc++.h>

using namespace std;

class Solution {
public:
    vector<int> findAnagrams(string s, string p) {
        if (s.size() < p.size()) return {};
        
        int m = s.size();
        int n = p.size();
        
        vector<int> hash(26), curr(26);
        for (auto c: p)
            hash[c-'a'] += 1;
        
        for (int i = 0; i < n; i++)
            curr[s[i]-'a'] += 1;
        
        vector<int> sol;
        if (hash == curr)
            sol.push_back(0);
        
        for (int i = n; i < m; i++) {
            curr[s[i-n]-'a'] -= 1;
            curr[s[i]-'a'] += 1;
            if (hash == curr)
                sol.push_back(i-n+1);
        }
        return sol;
    }
};

int main() {

    return 0;
}

File: algorithms/python/438-find-all-anagrams-in-a-string.py
# https://leetcode.com/problems/find-all-anagrams-in-a-string/

from typing import List

class Solution:
    def findAnagrams(self, s: str, p: str) -> List[int]:
        if len(s) < len(p): return []
        p_list = [ 0 for i in range(0, 26) ]
        s_list = [ 0 for i in range(0, 26) ]
        sol = []
        
        for i in range(0, len(p)):
            p_list[ord(p[i])-ord('a')] += 1
            s_list[ord(s[i])-ord('a')] += 1
        if p_list == s_list:
            sol.append(0)
            
        for i in range(len(p), len(s)):
            s_list[ord(s[i-len(p)])-ord('a')] -= 1
            s_list[ord(s[i])-ord('a')] += 1
            if p_list == s_list:
               sol.append(i-len(p)+1)
        return sol
        
